src/mlpack/methods/ann/layer/mean_pooling.hpp

```diff
@@ -238,98 +226,98 @@ class MeanPoolingType : public Layer<InputType, OutputType>
 
-          if (rowEnd > input.n_rows - 1)
+          if (rowEnd > output.n_rows - 1)
           {
             if (floor)
               continue;
-            rowEnd = input.n_rows - 1;
+            rowEnd = output.n_rows - 1;
           }
 
-          if (colEnd > input.n_cols - 1)
+          if (colEnd > output.n_cols - 1)
           {
             if (floor)
               continue;
-            colEnd = input.n_cols - 1;
+            colEnd = output.n_cols - 1;
           }
 
           size_t kernalArea = (rowEnd - i + 1) * (colEnd - j + 1);
           output(i, j) += error(rowidx, colidx) / kernalArea;
 
-          if (rowEnd + 1 < input.n_rows)
+          if (rowEnd + 1 < output.n_rows)
           {
             output(rowEnd + 1, j) -= error(rowidx, colidx) / kernalArea;
 
-            if (colEnd + 1 < input.n_cols)
+            if (colEnd + 1 < output.n_cols)
               output(rowEnd + 1, colEnd + 1) += error(rowidx, colidx) / kernalArea;
           }
 
-          if (colEnd + 1 < input.n_cols)
+          if (colEnd + 1 < output.n_cols)
             output(i, colEnd + 1) -= error(rowidx, colidx) / kernalArea;
         }
       }
 
-      for (size_t i = 1; i < input.n_rows; ++i)
+      for (size_t i = 1; i < output.n_rows; ++i)
         output.row(i) += output.row(i - 1);
 
-      for (size_t j = 1; j < input.n_cols; ++j)
+      for (size_t j = 1; j < output.n_cols; ++j)
         output.col(j) += output.col(j - 1);
     }
     else
     {
-      arma::Mat<eT> unpooledError;
-      for (size_t j = 0, colidx = 0; j < input.n_cols; j += strideHeight, ++colidx)
+      arma::Mat<typename MatType::elem_type> unpooledError;
+      for (size_t j = 0, colidx = 0; j < output.n_cols; j += strideHeight, ++colidx)
       {
-        for (size_t i = 0, rowidx = 0; i < input.n_rows; i += strideWidth, ++rowidx)
+        for (size_t i = 0, rowidx = 0; i < output.n_rows; i += strideWidth, ++rowidx)
         {
           size_t rowEnd = i + kernelWidth - 1;
           size_t colEnd = j + kernelHeight - 1;
 
-          if (rowEnd > input.n_rows - 1)
+          if (rowEnd > output.n_rows - 1)
           {
             if (floor)
               continue;
-            rowEnd = input.n_rows - 1;
+            rowEnd = output.n_rows - 1;
           }
 
-          if (colEnd > input.n_cols - 1)
+          if (colEnd > output.n_cols - 1)
           {
             if (floor)
               continue;
-            colEnd = input.n_cols - 1;
+            colEnd = output.n_cols - 1;
           }
 
-          arma::mat InputArea = input(arma::span(i, rowEnd), arma::span(j, colEnd));
+          arma::mat InputArea = output(arma::span(i, rowEnd), arma::span(j, colEnd));
 
-          unpooledError = arma::Mat<eT>(InputArea.n_rows, InputArea.n_cols);
+          unpooledError = arma::Mat<typename MatType::elem_type>(InputArea.n_rows, InputArea.n_cols);
           unpooledError.fill(error(rowidx, colidx) / InputArea.n_elem);
```

Comment:
Probably it would also make sense to move these functions into the `_impl.hpp` file, so that the header file does not get too cluttered.

Reply:
Done.